aws/s3.go

```diff
@@ -56,87 +65,197 @@ func getS3BucketTags(svc *s3.S3, bucketName *string) ([]map[string]string, error
 		}
 
-		if !excludeAfter.After(*bucket.CreationDate) {
-			logging.Logger.Debugf("Skipping - Bucket %s - matched CreationDate filter", *bucket.Name)
-			continue
+		for region, buckets := range currBucketNamesPerRegion {
+			if _, ok := bucketNamesPerRegion[region]; !ok {
+				bucketNamesPerRegion[region] = []*string{}
+			}
+			for _, bucket := range buckets {
+				bucketNamesPerRegion[region] = append(bucketNamesPerRegion[region], bucket)
+			}
 		}
+		batchCount++
+	}
+	return bucketNamesPerRegion, nil
+}
 
-		bucketRegion, err := getS3BucketRegion(svc, bucket.Name)
+// getRegions creates s3 clients for target regions
+func getRegionClients(regions []string) (map[string]*s3.S3, error) {
+	var regionClients = make(map[string]*s3.S3)
+	for _, region := range regions {
+		logging.Logger.Debugf("S3 - creating session - region %s", region)
+		awsSession, err := session.NewSession(&awsgo.Config{
+			Region: awsgo.String(region)},
+		)
 		if err != nil {
-			logging.Logger.Warnf("Skipping - Bucket %s - Failed to get bucket location.", *bucket.Name)
-			continue
+			return regionClients, err
 		}
+		regionClients[region] = s3.New(awsSession)
+	}
+	return regionClients, nil
+}
 
-		// Create session for current bucket region as bucket location's session is required for getting bucket tags.
-		if _, ok := regionClients[bucketRegion]; !ok {
-			logging.Logger.Debugf("Creating session - Bucket %s - region %s", *bucket.Name, bucketRegion)
-			awsSession, err := session.NewSession(&awsgo.Config{
-				Region: awsgo.String(bucketRegion)},
-			)
-			if err != nil {
-				return bucketNamesPerRegion, errors.WithStackTrace(err)
-			}
-			regionClients[bucketRegion] = s3.New(awsSession)
+// getBucketNamesPerRegions gets valid bucket names concurrently from list of target buckets
+func getBucketNamesPerRegion(svc *s3.S3, targetBuckets []*s3.Bucket, excludeAfter time.Time, regionClients map[string]*s3.S3,
+	bucketNameSubStr string) (map[string][]*string, error) {
+
+	var bucketNamesPerRegion = make(map[string][]*string)
+	var bucketCh = make(chan *S3Bucket, len(targetBuckets))
+	var wg sync.WaitGroup
+
+	for _, bucket := range targetBuckets {
+		if len(bucketNameSubStr) > 0 && !strings.Contains(*bucket.Name, bucketNameSubStr) {
+			logging.Logger.Debugf("Skipping - Bucket %s - failed substring filter - %s", *bucket.Name, bucketNameSubStr)
+			continue
 		}
 
-		validTags, excludeReason := hasValidTags(regionClients[bucketRegion], bucket)
-		if !validTags {
-			logging.Logger.Debugf("Skipping - Bucket %s - %s", *bucket.Name, excludeReason)
+		wg.Add(1)
+		go func(bucket *s3.Bucket) {
+			defer wg.Done()
+			getBucketInfo(svc, bucket, excludeAfter, regionClients, bucketCh)
+		}(bucket)
+	}
+
+	go func() {
+		wg.Wait()
+		close(bucketCh)
+	}()
+
+	for bucketData := range bucketCh {
+		if bucketData.Error != nil {
+			logging.Logger.Warnf("Skipping - Bucket %s - region - %s - error: %s", bucketData.Name, bucketData.Region, bucketData.Error)
```

Comment:
Should errors be skipped? Or cause the entire nuke operation to fail due to an error?

Reply:
As per earlier conversation - https://github.com/gruntwork-io/cloud-nuke/pull/96#discussion_r386855853 - the aim was to delete as many buckets as we can and then return a deleteCount which calls out if the deletion was 100% successfull or not https://github.com/gruntwork-io/cloud-nuke/blob/f579288286105f472751718c257638a5329834a6/aws/s3_types.go#L48